File: Tellma/Controllers/DTO/PermissionsForClient.cs
using System.Collections.Generic;

namespace Tellma.Controllers.Dto
{
    public class PermissionsForClient
    {
        public PermissionsForClientViews Views { get; set; }

        public List<int> ReportIds { get; set; }

        public List<int> DashboardIds { get; set; }
    }

    /// <summary>
    /// This DTO carries permission information to the client so
    /// the client can adjust the UI accordingly, the string key
    /// in the dictionary represents the View
    /// </summary>
    public class PermissionsForClientViews : Dictionary<string, Dictionary<string, bool>>
    {
        // View -> Action -> True
    }
}

File: Tellma.Api.Web/ClientApp/src/app/data/dto/permissions-for-client.ts
export interface PermissionsForClient {
  Permissions: UserPermission[];
  Views?: PermissionsForClientViews; // Added on the client side
  ReportIds: number[];
  DashboardIds: number[];
  TemplateIds: number[];
}

export interface UserPermission {
  View: string;
  Action: string;
  Criteria: string;
}

export interface PermissionsForClientViews {
  [view: string]: {
    [action: string]: boolean;
  };
}
